From 372b9e91d8984e8cc535ff69f2756ada88a87c64 Mon Sep 17 00:00:00 2001
From: JulienBolland <julien.bolland@student.uliege.be>
Date: Mon, 13 Apr 2020 22:28:58 +0200
Subject: [PATCH] print options

---
 kvs/kvs.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/kvs/kvs.c b/kvs/kvs.c
index 5f9c804..25f99d9 100644
--- a/kvs/kvs.c
+++ b/kvs/kvs.c
@@ -2,14 +2,19 @@
 
 asmlinkage long sys_kvs_insert(const char *key, size_t keylen,
 			       const char *val, size_t vallen){
-
+	printk("In insert\n");
+	return 0;
 }
 
 asmlinkage long sys_kvs_search(const char *key, size_t keylen,
                                char *val, size_t vallen){
+	printk("In search\n");
+        return 0;
 
 }
 
 asmlinkage long sys_kvs_delete(const char *key, size_t keylen){
+	printk("In delete\n");
+        return 0;
 
 }
-- 
2.7.4

